builder/evaluator.go

```diff
@@ -308,7 +311,14 @@ func (b *Builder) dispatch(stepN int, ast *parser.Node) error {
 		str = ast.Value
 		if _, ok := replaceEnvAllowed[cmd]; ok {
 			var err error
-			str, err = ProcessWord(ast.Value, b.Config.Env)
+			// Append the build-time environment to config-environment.
+			// This allows builder config to override the variables, making the behavior similar to
+			// a shell script i.e. `ENV foo bar` overrides value of `foo` passed in build
+			// context. But `ENV foo $foo` will use the value from build context if one
+			// isn't already been defined by a previous ENV primitive.
```

Comment:
// Note, we get this behavior because we know that `ProcessWord()` will
// stop on the first occurrence of a variable name and not notice
// a subsequent one. So, putting the BuildEnv list after the Config.Env 
// list, in 'envs', is safe.